
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/proto/dataplane/sai";

enum VirtualRouterAttr {
  VIRTUAL_ROUTER_ATTR_UNSPECIFIED = 0;
  VIRTUAL_ROUTER_ATTR_ADMIN_V4_STATE = 1;
  VIRTUAL_ROUTER_ATTR_ADMIN_V6_STATE = 2;
  VIRTUAL_ROUTER_ATTR_SRC_MAC_ADDRESS = 3;
  VIRTUAL_ROUTER_ATTR_VIOLATION_TTL1_PACKET_ACTION = 4;
  VIRTUAL_ROUTER_ATTR_VIOLATION_IP_OPTIONS_PACKET_ACTION = 5;
  VIRTUAL_ROUTER_ATTR_UNKNOWN_L3_MULTICAST_PACKET_ACTION = 6;
  VIRTUAL_ROUTER_ATTR_LABEL = 7;
}

message CreateVirtualRouterRequest {
  uint64 switch = 1;

  bool admin_v4_state = 2;
  bool admin_v6_state = 3;
  bytes src_mac_address = 4;
  PacketAction violation_ttl1_packet_action = 5;
  PacketAction violation_ip_options_packet_action = 6;
  PacketAction unknown_l3_multicast_packet_action = 7;
  bytes label = 8;
}

message CreateVirtualRouterResponse {
  uint64 oid = 1;
}

message RemoveVirtualRouterRequest {
  uint64 oid = 1;
}

message RemoveVirtualRouterResponse {}

message SetVirtualRouterAttributeRequest {
  uint64 oid = 1;
  oneof attr {
    bool admin_v4_state = 2;
    bool admin_v6_state = 3;
    bytes src_mac_address = 4;
    PacketAction violation_ttl1_packet_action = 5;
    PacketAction violation_ip_options_packet_action = 6;
    PacketAction unknown_l3_multicast_packet_action = 7;
    bytes label = 8;
  }
}

message SetVirtualRouterAttributeResponse {}

message GetVirtualRouterAttributeRequest {
  uint64 oid = 1;
  VirtualRouterAttr attr_type = 2;
}

message GetVirtualRouterAttributeResponse {
  VirtualRouterAttribute attr = 1;
}

service VirtualRouter {
  rpc CreateVirtualRouter(CreateVirtualRouterRequest)
      returns (CreateVirtualRouterResponse) {}
  rpc RemoveVirtualRouter(RemoveVirtualRouterRequest)
      returns (RemoveVirtualRouterResponse) {}
  rpc SetVirtualRouterAttribute(SetVirtualRouterAttributeRequest)
      returns (SetVirtualRouterAttributeResponse) {}
  rpc GetVirtualRouterAttribute(GetVirtualRouterAttributeRequest)
      returns (GetVirtualRouterAttributeResponse) {}
}
